### PR TITLE
  Detailed error responses when missing fields.

### PR DESCRIPTION
**Pull request by vqtran**
Refactored the QueryResult object to be truer to its name.
Querier now does a validation step that walks over a QueryRequest and
checks if all expected fields are there. If not, it makes Alexa announce
a useful error message.

Next step re handling errors:
  - Make these error responses interactive. Alexa will try to
    interactively fix invalid requests with you.
  - Fix up SchemaInferrer / errors where column/table doesn't exist.
    For example, right now:

    how many sales where product is speakers and counts are nine

    fails because SchemaInferrer tries to look up counts but it doesn't
    exist, only count does. This null pointers, when it really should
    propagate some error upwards back to the user.

### CODE REVIEW

**Comment by vqtran** [comment]
Also this isn't meant to be interactive yet, these changes felt big enough for one PR

File: src/main/java/echoquery/intents/AggregationHandler.java [hunk 1]
```diff
@@ -34,12 +34,6 @@ public SpeechletResponse respond(Intent intent, Session session) {
    * Exposed for testing purposes - SpeechletResponse is impossible to inspect.
    */
   public String getResponseInEnglish(Intent intent, Session session) {
-    try {
-      QueryResult result = querier.execute(QueryRequest.of(intent));
-      return result.toEnglish();
-    } catch (SQLException e) {
```
Comment (GabeLoins) [inline src/main/java/echoquery/intents/AggregationHandler.java]: I will miss runtime exceptions being an integral component of our basic error handling

File: src/main/java/echoquery/sql/Querier.java [hunk 2]
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
...
+      return Optional.of(new QueryResult(Status.FAILURE,
+          "I'm not sure what table you're interested in, or maybe you referred "
+          + "to one that doesn't exist? Please try again."));
+    }
+
+    if (!request.getAggregationFunc().isPresent()) {
+      return Optional.of(new QueryResult(Status.FAILURE,
+          "I can't tell if you want to know the number of rows or the average, "
+          + "sum, min, or max of a particular column in the "
+          + request.getFromTable().get() + " table. Please try again."));
```
Comment (GabeLoins) [inline src/main/java/echoquery/sql/Querier.java]: should we just default to count in this case? just a thought

File: src/main/java/echoquery/sql/Querier.java [hunk 3]
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
...
+      }
+      if (i < request.getComparisonColumns().size() - 1
+          && (i >= request.getComparisonBinaryOperators().size()
+              || !request.getComparisonBinaryOperators().get(i).isPresent())) {
+        return Optional.of(new QueryResult(Status.FAILURE,
+            "It seems like you have more than one where condition but no and "
+            + "or or connecting them, and I'm not sure what to do with this."
+            + " Please try again."));
+      }
+    }
```
Comment (GabeLoins) [inline src/main/java/echoquery/sql/Querier.java]: MEGAMETHOD

File: src/main/java/echoquery/sql/Querier.java [hunk 4]
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
     this.inferrer = new SchemaInferrer(conn);
   }
 
-  public QueryResult execute(QueryRequest request) throws SQLException {
+  public QueryResult execute(QueryRequest request) {
+    // First validate the request, interactively mending malformed requests.
+    Optional<QueryResult> invalidation = validate(request);
```
Comment (GabeLoins) [inline src/main/java/echoquery/sql/Querier.java]: should validate be in query request?

File: src/main/java/echoquery/sql/Querier.java (hunk 2 above)
Comment (vqtran) [inline src/main/java/echoquery/sql/Querier.java]: this sounded weird at first but i think it makes sense, yeah

**Comment by GabeLoins** [comment]
looks good !

**Comment by GabeLoins** [comment]
you can merge

File: src/main/java/echoquery/sql/Querier.java (hunk 3 above)
Comment (vqtran) [inline src/main/java/echoquery/sql/Querier.java]: i'll helper method this, although helper methods without reuse purpose is kind of meh

File: src/main/java/echoquery/sql/Querier.java (hunk 4 above)
Comment (vqtran) [inline src/main/java/echoquery/sql/Querier.java]: I think having a validator validating requests makes more sense than every request validating itself, more of a pipeline

File: src/main/java/echoquery/intents/AggregationHandler.java (hunk 1 above)
Comment (vqtran) [inline src/main/java/echoquery/intents/AggregationHandler.java]: Now it'll just be a sneaky component!

File: src/main/java/echoquery/sql/Querier.java (hunk 3 above)
Comment (GabeLoins) [inline src/main/java/echoquery/sql/Querier.java]: please dont I think its pretty readable right now I was just messing around

**Comment by vqtran** [comment]
Changed default to count, added comments for QueryRequest, Querier, QueryResult since I think they're pretty stable now, refactored tests to cut out like 200 lines

**Pull request merged**